Review#1

mlavrinenko opened the pull request:
Review#1

melnikov-m-i committed:
test

melnikov-m-i committed:
Сделал task1.

melnikov-m-i committed:
изменил структуру файлов

melnikov-m-i committed:
Сделал task2.

melnikov-m-i committed:
Сделал task3

melnikov-m-i committed:
попытка работы с chart.js

melnikov-m-i committed:
Рабочий вариант task4

mlavrinenko commented on task1/task1.js:
```diff
@@ -0,0 +1,34 @@
+    var tfWidth = document.getElementById('width');
+    var tfHeight = document.getElementById('height');
+    var elSquare = document.querySelector('.square');
+    var btColor = document.querySelector('.bt-random-color');
```
Переменную, которая используется лишь один раз стоит заводить только в случае если это существенно упрощает читаемость записи. В данном случае это не так. Лучше просто в нужном месте написать `document.querySelector('.bt-random-color')`

mlavrinenko commented on task1/task1.js:
```diff
@@ -0,0 +1,34 @@
+    var elSquare = document.querySelector('.square');
+    var btColor = document.querySelector('.bt-random-color');
+
+    btColor.onclick = function (event) {
```
onclick и другие аналогичные свойства - работают, но использовать их не стоит, т.к. элемент может иметь несколько обработчиков и если кто-то добавит свой обработчик таким же способом, то ты таким кодов затрешь чужой обработчик или же наоборот

addEventListener - наше все (https://www.w3schools.com/jsref/met_element_addeventlistener.asp)

mlavrinenko replied:
Это конечно касается и остального кода

mlavrinenko commented on task1/task1.js:
```diff
@@ -0,0 +1,34 @@
+            b: Math.random() * 255
+        };
+
+        elSquare.style.backgroundColor = "rgb(" + randomColor.r + ", " + randomColor.g + ", " + randomColor.b + ")";
```
Это конечно работает, но я бы не рекомендовал путать типы.
В данном случае конечный результат будет типа такого:
`rgb(172.24016894162378, 86.45178024713299, 123.02036875264238)`

Хотя в дробной части нет никакого смысла. Приведи к int'у - банальная аккуратность

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+
+    btSubmit.onclick = function (event) {
+        var fieldError = document.querySelector('.error-form');
+        var error = validateForm();
```
Логичнее было бы назвать errors. От error ожидаешь что это будет одной ошибкой, а на деле это не так

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+            var str = '';
+            error.forEach(function(item){
+                str += item + '<br>';
+            });
```
Отдельную переменную str заводить вовсе не нужно
`error.join('<br />')`

И к слову, вообще не рекомендую называть так переменные общими названиями. Если бы она была нужна, можно было бы назвать errorsStr хотя бы уж

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+            fieldError.innerHTML = str;
+        }
+        else {
+            fieldError.innerHTML = "Заказ оформлен!";
```
Полю ошибки присваивается сообщение об успехе :-) Надо либо поле назвать по другому (например messageArea), либо сделать их два. Иначе путаница.

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+
+        if(form.elements['phone'].value === '') {
+            error.push('Не заполнено поле Телефон');
+        }
```
А теперь представь что в форме 100 полей из них 50 - обязательные.
Сколько однообразного кода придётся добавить?

В таких случаях (даже если всего два элемента), все таки напрашивается цикл.
Имя поля можешь взять из placeholder-атрибута

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+            error.push('Телефон должен содержать только числа');
+        }
+
+        if(form.elements['email'].value !== '' && form.elements['email'].value.indexOf('@') == -1) {
```
Для пустой строки последнее условие тоже не сработает, поэтому первое условие излишне

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+    ymaps.ready(init);
+    function init(){
+        var myMap = new ymaps.Map("map", {
+            center: [54.1941, 37.6139],
```
Желательно такие вещи выносить в константы.
Чтобы было очевидно - а куда ты хотел установить центр?

`let moscowCoords = [54.1941, 37.6139];`
`center: moscowCoords`
- в итоге читается бодрее

В данном случае так надо было сделать потому что понятно что цифры означают центр, но центр чего?

Например в случае `zoom: 12` магии особо нет, потому что мы видим что означает это цифра из-за того куда она попадает. Но это все ещё хардкод. Который впрочем здесь позволителен

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,74 @@
+            });
+
+            myMap.geoObjects.removeAll();
+            myMap.geoObjects.add(myPlacemark);
```
А можно как-то избежать removeAll?
Ну, скажем если бы была задача выводить 1000 точек ПВЗ и была бы возможность добавить новую точку, то удалять весь список и добавлять все точки заново - процедура излишняя.

Понятное дело - в данном примере это ни разу не важно с точки зрения производительности, но все таки 
с точки зрения алгоритма оптимальнее было бы либо удалить ту точку которая была на предыдущем шаге добавлена и создать новую, либо, ещё лучше - изменить параметры существующей точки.

mlavrinenko commented on task3/task3.js:
```diff
@@ -0,0 +1,95 @@
+    });
+
+    btUpdate.onclick = function(event) {
+        getDataOfProducts().then(function(data) {
```
В задании сказано "Сервис возвращает одни и те же данные в течение дня"
Здесь же мы при каждой фильтрации получаем данные снова. Зачем?

mlavrinenko commented on task3/task3.js:
```diff
@@ -0,0 +1,95 @@
+
+    getDataOfProducts().then(function(data) {
+        generateTable(data);
+    });
```
Можно записать короче: `getDataOfProducts().then(generateTable)`

mlavrinenko commented on task3/task3.js:
```diff
@@ -0,0 +1,95 @@
+                } else {
+                    resultFilter = data.filter(function(product) {
+                        return product.price >= fieldFromPrice.value && product.price <= fieldToPrice.value;
+                    });
```
Попробуй реализовать фильтрацию с помощью одной else ветки

mlavrinenko commented on task3/task3.js:
```diff
@@ -0,0 +1,95 @@
+                str += '<td>' + products[i].quantity * products[i].price + '</td>';
+                str += '</tr>';
+            }
+            str += '</tbody></table>';
```
Генерация HTML через строки - так себе выглядит и читается. Попробуй найти альтернативу.
К слову, это касается и PHP. Если где-то будешь генерить HTML склеивая строки, на ревью с большой долей вероятности возникнут вопросы

mlavrinenko commented on task4/task4.js:
```diff
@@ -0,0 +1,119 @@
+                    ". Процент запросов, вернувших ошибку: " + (countRequestOfError/countRequest*100).toPrecision(2) + "%";
+            }
+        });
+        setTimeout(go, 5000);
```
Если запрос к серверу подвиснет более чем на 5сек, то может получиться так, что запросы выполнятся считай одновременно.

Если мы не должны ждать завершения обработки предыдущего запроса, то можно просто повести функцию через setInterval
Если же нам надо ждать завершения обработки, то надо разместить код внутри then.

Давай в качестве обучения сделаем именно последний вариант.
Обрати внимание что у тебя есть ветка с return.
Правильнее всего было бы вернуть в then промис, из которого ты будешь вызывать resolve перед return, а так же в конце регулярного кода.
В последующем then уже можно вызвать setTimeout

melnikov-m-i committed:
сделал динамическое добавленние (удаление) последней строки (столбца), редактирование ячейки без сохранения в localStorage, начал работу с localStorage.

melnikov-m-i committed:
Сделал так что при редактировании ячейки, создании и удалении строк(столбцов) данные записывались(добавлялись, удалялись) в localStorage.
Также сделал возможность редактировании при двойном тапе. Поправил стили.

melnikov-m-i commented:
Внёс указанные правки в task1-task4, commit 4d53128722acc737a5449d78d3a1fcd9bbbd0f97

melnikov-m-i committed:
Внёс указанные правки в task1-task4

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,79 @@
+    var errors = [];
+
+    for(let i = 0; i < form.elements.length; i++) {
+        if(form.elements[i].required) {
```
Если мы показываем что поле не заполнено, то показывать ошибки проверки по формату уже не нужно.
Кажется, это не наш случай, но сегодня поле не обязательное, а завтра обязательное, бывает.

Добавь email'у обязательность и попробуй решить эту проблему.
Классическое решение - заведение ассоциативного массива который описывает логику проверки каждого поля

```javascript
var validators = {
   email: {
      required: true,
      validate: function (email) {
             return email.indexOf('@') > 0;
      }
   }
};
```
Опиши остальные поля и переработай код под этот массив, так чтобы про одно поле не было две ошибки (про обязательность заполнение и формат)

mlavrinenko commented on task2/task2.js:
```diff
@@ -0,0 +1,79 @@
+            myPlacemark.properties.set({
+                balloonContentBody: '<p>Координаты доставки: ' + strCoords + '</p>',
+                hintContent: strCoords
+            });
```
Уже лучше, но теперь появилось дублирование кода. Как минимум может поменяться текст "Координаты доставки" и придется делать правку в двух местах - не хорошо.

В данном случае я бы до if/else завел ассоциативный массив с минимальным набором полей который точно будет нужен (в данном случае balloonContentBody и hintContent), а в if-ветке его дополнил бы полем balloonContentHeader.

В обоих ветках вместо inline-массива будет передаваться эта переменная

mlavrinenko commented on task5/task5.js:
```diff
@@ -0,0 +1,369 @@
+    table.insertButtonsForRows();
+    table.insertButtonsForColumns();
+
+    table.getTable().addEventListener('dblclick', table.editCell.bind(table));
```
table.getTable - выглядит странновато.
Это достаточно распространённая тема - упоминать в методах имя класса или что-то очевидное.
Лучше этого избегать.

В данном случае лучше назвать getNode. И т.к. ты дважды вызываешь к нему (table.getTable()), то лучше вынести в переменную, т.е.:

```javascript
var table = new Table('.container'),
    tableNode = table.getNode();
```

mlavrinenko commented on task5/task5.js:
```diff
@@ -0,0 +1,369 @@
+            return false;
+        } else {
+            let index = table.rows.length;
+            let dataLS = localStorage.getItem(this._getKeyLocalStorage());
```
Эта операция здесь и в других подобных методах в общем-то лишняя.
Представь что из-за какого-то глюка произойдёт рассинхрон между отображаемыми данными и хранилищем.

Наиболее логичная схема - компонент изначально считывает данные с хранилища и сохраняет их у себя в переменной. По этой переменной производится генерация данных для компонента.
Когда нужно изменить значения, меняем данные в переменной и перезапускаем рендер. 
Так работает, например, React.

Но, здесь конечно так усложнять не надо. Достаточно изначально считать данные и модифицировать их синхронной с изменениями на форме и сохранять в хранилище. А считывания достаточно одного.

mlavrinenko commented on task5/task5.js:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    _startLocalStorage() {
+        this.keyLocalStorage = 'el-table-' + Math.random().toString(36).substr(2, 9);
```
Я что-то если честно не совсем понял в чем смысл - просто так в хранилище данные закидывать?
Ну, вероятно в задании по сути так и написано, но скорее всего имелось ввиду другое - надо иметь возможность обновить страницу и продолжить редактирование в том же виде как до обновления страницы

mlavrinenko commented on task5/task5.js:
```diff
@@ -0,0 +1,369 @@
+
+        td.insertAdjacentHTML("beforeEnd",
+            '<div class="edit-controls"><button class="edit-ok">OK</button><button class="edit-cancel">CANCEL</button></div>'
+        );
```
OK/Cancel все таки лишние, это не особо удобно
Лучше сохранять ячейку при потере фокуса
Но при нажатии Esc мы отменяем изменение

И надо бы как-то сделать так, чтобы ячейка не меняла размеры при всем при этом.

mlavrinenko commented on task5/task5.js:
```diff
@@ -0,0 +1,369 @@
+        return arTable;
+    }
+
+    _writeDataInLocalStorage(data) {
```
Отмечу что после того как данные будут храниться в переменной объекта можно будет сделать просто `save()`

melnikov-m-i committed:
правка недочетов

mlavrinenko commented on task2/task2.js:
```diff
@@ -16,26 +17,60 @@ function ready() {
+                errors.push(validators[fieldName].messageNoValidation);
+            }
+        } else if (fieldName === 'map') {
+            if (!validators[fieldName].validate(window.myPlacemark)) {
```
Глобальных переменных лучше избегать.
В данном случае достаточно было бы завести переменную над функцией, а весь код файла обрамить в конструкцию типа:

```javascript
(function () {
    // код
})();
```

Этот классический прием нацелен на то чтобы избежать объявления глобальных переменных для кода, потому что функция будет иметь свою область видимости.

Дополню, что в проектах со сборкой кода (у нас обычно webpack), где код организован модульно, так уже делать не нужно, т.к. там все переменные объявляются уже в локальной видимости данного файла. Впрочем, это уже другая история.

mlavrinenko commented on task5/task5.js:
```diff
@@ -264,39 +230,31 @@ class Table {
-        let editControls = this.table.querySelector('.edit-controls');
+        input.addEventListener('keydown', function (e) {
+            e = e || window.event;
+            if (e.code == 'Escape' || e.key == 'Escape' || e.keyCode == 27) {
```
Убрать магическое число.

Такой пример самым очевидным образом показывает суть проблемы. Может быть у тебя сейчас в голове есть понимание того что 27 это Escape, но вот через пару недель его может и не быть. Как и у других разработчиков которые будут читать такой код.

Лучше заводить на такие случаи Enum-объекты и использовать уже их:

```javascript
const KEYCODE = {
    ESC: 27
};
```

Отмечу, что опять же, если в проекте будет использована сборка, легче подключить какой-то готовый пакет, где это все уже объявлено (к примеру сходу находится https://www.npmjs.com/package/keycode-js)

melnikov-m-i committed:
Правка недочетов:
в task2 замена объявления глобальной переменной на локальную переменную функции мгновенного вызова, в task5 замена магического числа константу, решение проблемы срабатывания blur при нажатии esc.
